### ISSUE TITLE
`separators` for json_format.MessageToJson

### ISSUE THREAD

**Issue by Clement-Jean** [protocolbuffers/protobuf#9659]
**What language does this apply to?**
This is concerning the python `google.protobuf.json_format`

**Describe the problem you are trying to solve.**
I'm trying to compress the JSON string that is resulting from json_format.MessageToJson, I've done something like:

```python
json_format.MessageToJson(message, indent=None)
```

However there are still some insignificant whitespaces. The code given above give a string like:

```json
{"id": 42, "isSimple": true, "name": "My name", "sampleLists": [3, 4, 5]}
```

what I want is:

```json
{"id":42,"isSimple":true,"name":"My name","sampleLists":[3,4,5]}
```

**Describe the solution you'd like**
In order to stay consistent with the Go `google.golang.org/protobuf/encoding/protojson` and the Java `com.google.protobuf.util.JsonFormat`, it would be great if we can add the named parameter called separators to MessageToJson and pass that to the `json.dumps` function in the `_Printer.ToJsonString`. We could also add the have a default value of:

```python
separators=(', ', ': ')
```

So that it is directly backward compatible.

**Describe alternatives you've considered**
The alternatives I considered would be:
- to trim the whitespaces manually, which is not really necessary since `json.dumps` already support that.
- change the `indent` interpretation, but that would make the behavior non consistent with `json.dump` (which I think is important, to keep)

**Additional context**

I would like to get an approval to work on this as a Pull Request, I read [here](https://github.com/protocolbuffers/protobuf/blob/master/CONTRIBUTING.md#pull-request-guidelines) that it would be better to discuss about it here first.

**Comment by elharo** [comment]
The question here is whether the benefit outweighs the extra complexity and support burden in the API. Is there a reason you need to remove the whitespace in the output? It makes the output less readable for a trivial space savings. Since we already have this in other languages, I suppose it might be OK. 

I don't think a list of separators is the way to go though. This should be a simple boolean consistent with the other languages.

**Comment by Clement-Jean** [comment]
Thanks for the feedback, the reason I want that is because on a large JSON the space saving is not trivial. I just gave a small example to explain what I was trying to do.

Then I thought about the separator list just to be consistent with the `json.dumps` function, the `ToJsonString` already use parameters that are similarly named and typed. But I guess a Boolean would also be fine.

I kind of like the Java way of doing that (`omittingInsignificantWhitespace`), but I guess a more pythonic way of doing that would be to add a parameter like `trim_spaces` or `trim_whitespaces`. Any feedback on that is welcome

**Comment by elharo** [comment]
How much savings do you estimate? How does that compare to simply gzipping the file?

**Comment by Clement-Jean** [comment]
You are probably right for the gzip part, did a little experiment on a 196K JSON and I got the following:

```
196K  default-widths.json
173K  default-widths_half_compressed.json
35K   default-widths_half_compressed.json.gzip
150K default-widths_compressed.json
35K   default-widths_compressed.json.gzip
```

where the gzip + 'half compressed' (understand with insignificant spaces) and gzip + compressed are finally both at 35K.

**Comment by haberman** [comment]
JSON format is not at all optimized for size. If size is a concern, binary format is a much better choice.

If you are size-conscious, why not just use binary format?

**Comment by Clement-Jean** [comment]
I'm totally aware of that. But I see 2 potential reasons why I would use JSON. Here they are:

- Debugging. This would be feasible for small size JSON and it wouldn't clutter my terminal. However I agree that pretty printed would be OK too.

- Compatibility with older apps. If my/others older code is only accepting JSON and I do not have time to change that or I do not have control over the project, I would like to have a payload as small as possible.

I agree that compressing the original JSON covers most of the payload size problem. We could still go further and remove the unnecessary spaces. But I also understand the problem in terms of maintaining the project.

**Comment by haberman** [comment]
Yes, we generally try to keep roughly the same set of options between languages, so adding it to Python would imply adding it to other languages as well. For this reason we try to keep to the set of options with the biggest impact.

One option would be to parse and re-serialize the JSON if you want it to be in a specific format. That's probably not an entirely satisfying answer, as it will have some performance cost, but it's something you can do today.
